app/src/main/java/org/dicio/dicio_android/eval/SkillEvaluator.java

```diff
@@ -353,14 +372,40 @@ private void evaluateMatchingSkill(final String input) {
             final List<String> inputWords = WordExtractor.extractWords(input);
             final List<String> normalizedWordKeys = WordExtractor.normalizeWords(inputWords);
             final Skill skill = skillRanker.getBest(input, inputWords, normalizedWordKeys);
-
+            if(skill==null) {
+                Log.w("SKILLS", "No matching Skill found");
+                return null;
+            }
             skill.processInput(SkillHandler.getSkillContext());
             return skill;
         })
         .subscribeOn(Schedulers.io())
         .observeOn(AndroidSchedulers.mainThread())
         .subscribe(this::generateOutput, this::onError);
     }
+    private void evaluateMatchingSkill(final String[] inputs) {
+        if (evaluationDisposable != null && !evaluationDisposable.isDisposed()) {
+            evaluationDisposable.dispose();
+        }
+
+        evaluationDisposable = Single.fromCallable(() -> {
+            Skill skill=null;
+            for(String input :inputs){
+                final List<String> inputWords = WordExtractor.extractWords(input);
+                final List<String> normalizedWordKeys = WordExtractor.normalizeWords(inputWords);
+                skill = skillRanker.getBest(input, inputWords, normalizedWordKeys);
+                if (skill != null&&!(skill instanceof TextFallback)) {
+                    break;
+                }
```

Comment:
Instead of putting the loop here, I'd delegate it to `skillRanker.getBest`. So the signature for (the only) `getBest` would be `getBest(List<String>, List<List<String>>, List<List<String>>)` This would prevent duplicating the code for `evaluateMatchingSkill`.